### ISSUE TITLE
[Discuss] Support providing hints for breaking changes

### ISSUE THREAD

**Issue by jsoriano** [elastic/package-spec#817]
There are situations where breaking changes may be unavoidable. Currently packages can indicate this in changelogs and/or bumping their major versions, but from a user perspective upgrading to versions with breaking changes looks the same as any other upgrade. After the upgrade they may be missing fields or having other kinds of issues.

It should be possible to include hints in packages to handle breaking changes. There can be two kinds of hints:
* Hints for Fleet, in case it can do anything during the upgrade.
* Hints for users, in case they need to be aware of some change or perform some manual action.

These hints should help with the migration path to versions with breaking changes.

There are some places where the hints could appear:
* In breaking change notes in the changelog, a new field could be added with additional information. We should ensure that breaking changes are displayed to users in the UI when upgrading (https://github.com/elastic/kibana/issues/187481), and if possible also in API responses.
* In field definitions, we could keep definitions of removed fields, along with additional information to indicate if it was completely removed, moved or whatsoever.
* In READMEs we could also add some notes, but this can be difficult to enforce and users may not read it.

We should also assess frequent reasons for breaking changes.

Try to find opportunities to improve the situation here.

Related issues
---

* https://github.com/elastic/package-spec/issues/229
* https://github.com/elastic/kibana/issues/187481
* https://github.com/elastic/elastic-package/issues/579. In this issue we help preventing breaking changes, but what to do next after detecting a breaking change if the change cannot be avoided?
* https://github.com/elastic/package-spec/issues/192 and https://github.com/elastic/package-spec/issues/227 contain discussions about deprecations, an special case of breaking changes.

**Comment by Oddly** [comment]
Not sure if this is the right place, but deprecation hints when using elastic-package to build packages would be very welcome for those of us who build our own packages.

**Comment by jlind23** [comment]
@kpollich @jamiehynds @nimarezainia instead of over engineering this could we imagine an easy solution which would be to display a pop up warning the users whenever there is major version bumped in the package ready to be installed? This would avoid us to implement any flags or any logic in the package spec itself. Any thoughts?

**Comment by kpollich** [comment]
Do we enforce a major version bump in the integration's version when its targeted stack/Kibana version is bumped by a major? I think this could be a simple way to get some of this behavior, but I don't know that we have any kind of enforcement on package maintainers that would guarantee we trigger the pop-up when it's intended.

**Comment by ebeahan** [comment]
>  I don't know that we have any kind of enforcement on package maintainers that would guarantee we trigger the pop-up when it's intended.

The changelog spec includes a [`breaking-change` type ](https://github.com/elastic/package-spec/blob/main/spec/integration/changelog.spec.yml#L30)maintainers should use for any breaking changes.

Maybe the upgrade banner includes a warning if a newer changelog entry contains a `breaking-change`?

<img width="810" alt="Image" src="https://github.com/user-attachments/assets/488d1a2e-ee86-4da5-8643-27de4728ded6" />

**Comment by kpollich** [comment]
I think Eric's suggestion is the best path forward here. We can detect when the upgrade from current version to latest version includes a breaking change by parsing the `changelog.yml`. Then we can display a warning in this upgrade callout + the breaking change description as it appears in `changelog.yml`.

**Comment by jamiehynds** [comment]
Could we take this a step further by not only providing a hint about a breaking change but also requiring users to acknowledge it before proceeding with an update? The need for acknowledgment could be based on the severity of the change, as determined by PM and engineering.

For example, in [this case](https://github.com/elastic/integrations/pull/12916), it would be beneficial to clearly inform users about the impact of the breaking change and require explicit acknowledgment before they update. This approach ensures users are fully aware of the implications and prevents any unexpected disruptions.

**Comment by kpollich** [comment]
Moving this out of our iteration and will flesh out https://github.com/elastic/kibana/issues/187481 with some UX mockups to make it actionable.

**Comment by efd6** [comment]
An issue exists with the `changes.type` field in that it makes the implicit claim that breaking-change is orthogonal to both bugfix and enhancement since only one of the three possible options can be used in any given change, but this is not the case; a breaking change is necessarily either a bugfix or an enhancement, so it should be the case that you can set both. Ideally, it would have been a separate boolean field but that horse has already bolted, so it would be good if the type could be relaxed to support an array of string so that they could be treated as tags.

**Comment by jsoriano** [comment]
I remember someone [maybe you? 🤔] suggested once the introduction of a new type like `breaking-bugfix`, that would be both a bugfix and a breaking change. `breaking-change` would be used then for other changes, like enhancements, that are also breaking changes.

**Comment by efd6** [comment]
I don't think that was me. It would work, but it seems like a hack to work around the the type system here.
